### PR TITLE
HIVE-28015: Iceberg: Add identifier-field-ids support in Hive

### PR DESCRIPTION
**Pull request by zhangbutao**
### What changes were proposed in this pull request?

Support `create `iceberg table with identifier field using hive primary key syntax. Then other engine like Flink can do upsert( writing equality deletes) on the table created by hive.
`create table testprikeyice(id int primary key disable novalidate, name string) stored by iceberg;`

I will complete `alter `  add/drop primary key in the following task if we think this change is good.

BTW, in current hive code, we can not get primary key through `show create table ` statement, and this will make it inconvenient for users to check primary key in hive jdbc client(beeline). I think we can optimize the `show create table ` in following task.

### Does this PR introduce _any_ user-facing change?

No
### Is the change a dependency upgrade?

No

### How was this patch tested?

Added new UTs

### CODE REVIEW

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java [hunk 1]
```diff
@@ -1479,7 +1479,7 @@ public void createTable(CreateTableRequest request) throws
 
     HiveMetaHook hook = getHook(tbl);
     if (hook != null) {
-      hook.preCreateTable(tbl);
+      hook.preCreateTable(request);
```
[the remark this answers is not in the document]
Comment (zhangbutao) [reply standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java:1482]: If it is ok. maybe i also want to pass whole requet when executing `alter table ` operation.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 2]
```diff
@@ -194,6 +197,10 @@ public HiveIcebergMetaHook(Configuration conf) {
 
   @Override
   public void preCreateTable(org.apache.hadoop.hive.metastore.api.Table hmsTable) {
```
Comment (deniskuzZ) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:199]: is that an empty impl? that won't be a backward compatible change

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 3]
```diff
@@ -384,7 +398,7 @@ private void doPreAlterTable(org.apache.hadoop.hive.metastore.api.Table hmsTable
       preAlterTableProperties = new PreAlterTableProperties();
       preAlterTableProperties.tableLocation = sd.getLocation();
       preAlterTableProperties.format = sd.getInputFormat();
-      preAlterTableProperties.schema = schema(catalogProperties, hmsTable);
+      preAlterTableProperties.schema = schema(catalogProperties, hmsTable, null);
```
Comment (deniskuzZ) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:401]: maybe `Collections.emptySet()`

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 4]
```diff
@@ -803,12 +818,34 @@ private Schema schema(Properties properties, org.apache.hadoop.hive.metastore.ap
       // Add partitioning columns to the original column list before creating the Iceberg Schema
       List<FieldSchema> cols = Lists.newArrayList(hmsTable.getSd().getCols());
       cols.addAll(hmsTable.getPartitionKeys());
-      return HiveSchemaUtil.convert(cols, autoConversion);
+      Schema schema = HiveSchemaUtil.convert(cols, autoConversion);
+      return getSchemaWithIdentifierFields(schema, identifierFields);
```
Comment (deniskuzZ) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:822]: what is the diff between those 2 if blocks? can't we combine them
````
if (hmsTable.isSetPartitionKeys() {
     cols.addAll(hmsTable.getPartitionKeys());
}
````
rest is copy-paste

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 5]
```diff
@@ -803,12 +818,34 @@ private Schema schema(Properties properties, org.apache.hadoop.hive.metastore.ap
...
+    Set<Integer> identifierFieldIds = identifierFields.stream()
+            .map(column -> {
+              Types.NestedField field = schema.findField(column);
+              Preconditions.checkNotNull(field,
+                      "Cannot find identifier field ID for the column %s in schema %s", column, schema);
+              return field.fieldId();
+            })
+            .collect(Collectors.toSet());
+
+    List<Types.NestedField> cols = schema.columns().stream()
```
Comment (deniskuzZ) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:842]: why do we need 2 scans?
````
List<Types.NestedField> cols = Lists.transform(schema.columns(),
        column -> identifierFields.contains(column.name()) ? column.asRequired() : column);
````

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 6]
```diff
@@ -234,7 +241,12 @@ public void preCreateTable(org.apache.hadoop.hive.metastore.api.Table hmsTable)
     // - InputFormatConfig.TABLE_SCHEMA, InputFormatConfig.PARTITION_SPEC takes precedence so the user can override the
     // Iceberg schema and specification generated by the code
 
-    Schema schema = schema(catalogProperties, hmsTable);
+    Set<String> identifierFields = Optional.ofNullable(request.getPrimaryKeys())
```
Comment (deniskuzZ) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:244]: why do we need Set here?

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (deniskuzZ) [reply standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java:1482]: 👍

**Comment by deniskuzZ** [comment]
@zhangbutao, PR looks good, some minor stuff

**Comment by zhangbutao** [comment]
> @zhangbutao, PR looks good, some minor stuff 

Thanks for your review! I've been out recently. Will come back asap. 😀

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java (hunk 3 above)
Comment (zhangbutao) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java]: Fixed.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java (hunk 4 above)
Comment (zhangbutao) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java]: Simplified this code block.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java (hunk 2 above)
Comment (zhangbutao) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:199]: Fixed.

**Review by deniskuzZ — APPROVED** [review]
+1

**Pull request merged**